libsolidity/lsp/LanguageServer.cpp

```diff
@@ -0,0 +1,955 @@
+					m_remappings.emplace_back(move(remappingOpt.value()));
+				else
+					trace("Failed to parse remapping: '"s + element.asString() + "'");
+			}
+		}
+	}
+}
+
+namespace { // TODO: move us up!
+	inline size_t toOffset(std::string const& _text, langutil::LineColumn _position) noexcept
+	{
+		// TODO: take care of Unicode.
+		size_t offset = 0;
+		langutil::LineColumn current = {0, 0};
+		while (current != _position && offset < _text.size())
+		{
+			if (_text.at(offset) != '\n')
+				current.column++;
+			else
+			{
+				current.line++;
+				current.column = 0;
+			}
+			offset++;
+		}
+		return offset;
+	}
+
+	inline std::pair<size_t, size_t> offsetsOf(std::string const& _text, LineColumnRange _range) noexcept
+	{
+		return std::pair{
+			toOffset(_text, _range.start),
+			toOffset(_text, _range.end)
+		};
+	}
+}
+
+void LanguageServer::documentContentUpdated(string const& _path, LineColumnRange _range, string const& _replacementText)
+{
+	// TODO: all this info is actually unrelated to solidity/lsp specifically except knowing that
+	// the file has updated, so we can  abstract that away and only do the re-validation here.
+	auto file = m_fileReader->sourceCodes().find(_path);
+	if (file == m_fileReader->sourceCodes().end())
+	{
+		log("LanguageServer: File to be modified not opened \"" + _path + "\"");
+		return;
+	}
+
+	auto& buffer = file->second;
+
+	auto const [start, end] = offsetsOf(buffer, _range);
+	buffer.replace(start, end - start, _replacementText);
+	buffer.replace(start, end - start, _replacementText);
+}
+
+void LanguageServer::documentContentUpdated(string const& _path, string const& _replacementText)
+{
+	auto file = m_fileReader->sourceCodes().find(_path);
+	if (file == m_fileReader->sourceCodes().end())
+	{
+		log("LanguageServer: File to be modified not opened \"" + _path + "\"");
+		return;
+	}
+
+	file->second = _replacementText;
+
+	validate(_path);
+}
+
+frontend::ReadCallback::Result LanguageServer::readFile(string const& _kind, string const& _path)
+{
+	return m_fileReader->readFile(_kind, _path);
+}
+
+constexpr DiagnosticSeverity toDiagnosticSeverity(Error::Type _errorType)
+{
+	using Type = Error::Type;
+	using Severity = DiagnosticSeverity;
+	switch (_errorType)
+	{
+		case Type::CodeGenerationError:
+		case Type::DeclarationError:
+		case Type::DocstringParsingError:
+		case Type::ParserError:
+		case Type::SyntaxError:
+		case Type::TypeError:
+			return Severity::Error;
+		case Type::Warning:
+			return Severity::Warning;
+	}
+	// Should never be reached.
+	return Severity::Error;
+}
+
+bool LanguageServer::compile(std::string const& _path)
+{
+	// TODO: optimize! do not recompile if nothing has changed (file(s) not flagged dirty).
+
+	auto const i = m_fileReader->sourceCodes().find(_path);
+	if (i == m_fileReader->sourceCodes().end())
+		return false;
+
+	m_compilerStack.reset();
+	m_compilerStack = make_unique<CompilerStack>(bind(&FileReader::readFile, ref(*m_fileReader), _1, _2));
```

Comment:
Do you really need the file reader here or can you just pass all known files as a map of strings?